File: api/app/interfaces/schemas/policy.py
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field

from app.domain.models.policy import Policy
from app.domain.models.policy_match import PolicyMatch


class PolicyListItem(BaseModel):
    """政策列表项(不含正文，保持列表轻量)"""
    id: str = ""
    source: str = ""
    source_url: str = ""
    index_number: str = ""
    title: str = ""
    issuer: str = ""
    doc_number: str = ""
    status: str = ""
    publish_date: Optional[date] = None
    region: str = ""

    @classmethod
    def from_domain(cls, p: Policy) -> "PolicyListItem":
        return cls(
            id=p.id, source=p.source, source_url=p.source_url,
            index_number=p.index_number, title=p.title, issuer=p.issuer,
            doc_number=p.doc_number, status=p.status,
            publish_date=p.publish_date, region=p.region,
        )


class PolicyListResponse(BaseModel):
    """政策分页列表响应"""
    items: List[PolicyListItem] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 20


class PolicyDetailResponse(PolicyListItem):
    """政策详情响应(含正文、时间戳与申报截止)"""
    body_text: str = ""
    crawled_at: Optional[datetime] = None
    # ---- 申报截止(主线⑤；LLM 抽取，以原文为准、供参考核对) ----
    apply_deadline: Optional[date] = None
    apply_window_text: str = ""
    deadline_status: str = "unknown"  # extracted / rolling / unknown

    @classmethod
    def from_domain(cls, p: Policy) -> "PolicyDetailResponse":
        return cls(
            id=p.id, source=p.source, source_url=p.source_url,
            index_number=p.index_number, title=p.title, issuer=p.issuer,
            doc_number=p.doc_number, status=p.status,
            publish_date=p.publish_date, region=p.region,
            body_text=p.body_text, crawled_at=p.crawled_at,
            apply_deadline=p.apply_deadline, apply_window_text=p.apply_window_text,
            deadline_status=p.deadline_status,
        )


class PolicyMatchItem(BaseModel):
    """单条政策匹配候选(③匹配输出)：政策概要 + 融合分 + 可解释理由"""
    policy: PolicyListItem  # 政策概要(列表轻量，正文经详情接口按需取)
    score: float = 0.0  # RRF 融合总分(越大越靠前)
    structured_score: float = 0.0  # 结构化命中归一化分∈[0,1]
    semantic_score: float = 0.0  # 语义最高相似度∈[-1,1]
    matched_terms: List[str] = Field(default_factory=list)  # 命中的档案词
    reasons: List[str] = Field(default_factory=list)  # 推荐理由

    @classmethod
    def from_domain(cls, m: PolicyMatch) -> "PolicyMatchItem":
        return cls(
            policy=PolicyListItem.from_domain(m.policy),
            score=m.score, structured_score=m.structured_score,
            semantic_score=m.semantic_score, matched_terms=m.matched_terms,
            reasons=m.reasons,
        )


class PolicyMatchResponse(BaseModel):
    """政策匹配响应：候选列表(已按融合分倒序)"""
    items: List[PolicyMatchItem] = Field(default_factory=list)
    total: int = 0


class PolicySourceItem(BaseModel):
    """一个公开政策来源(地区/门户)"""
    key: str = ""  # 稳定标识(抓取时传入)
    name: str = ""  # 展示名
    region: str = ""  # 适用地区


class PolicySourceListResponse(BaseModel):
    """可抓取的政策来源列表"""
    items: List[PolicySourceItem] = Field(default_factory=list)

File: ui/src/lib/api/policy.ts
import { get, post } from "./fetch";

// ==================== 公开政策库类型 ====================

/** 政策列表项（不含正文） */
export type PolicyListItem = {
  id: string;
  source: string;
  source_url: string;
  index_number: string;
  title: string;
  issuer: string;
  doc_number: string;
  status: string;
  publish_date: string | null;
  region: string;
};

/** 申报截止抽取状态（主线⑤） */
export type DeadlineStatus = "extracted" | "rolling" | "unknown";

/** 政策详情（含正文与申报截止） */
export type PolicyDetail = PolicyListItem & {
  body_text: string;
  crawled_at: string | null;
  // 申报截止（LLM 抽取，以原文为准、供参考核对）
  apply_deadline: string | null;
  apply_window_text: string;
  deadline_status: DeadlineStatus;
};

/** 政策分页列表响应 */
export type PolicyListResponse = {
  items: PolicyListItem[];
  total: number;
  page: number;
  page_size: number;
};

/** 列表查询参数 */
export type ListPoliciesParams = {
  page?: number;
  page_size?: number;
  region?: string;
  issuer?: string;
  keyword?: string;
};

/** 单条政策匹配候选（③匹配输出） */
export type PolicyMatchItem = {
  policy: PolicyListItem;
  score: number; // RRF 融合总分
  structured_score: number; // 结构化命中归一化分 [0,1]
  semantic_score: number; // 语义最高相似度 [-1,1]
  matched_terms: string[]; // 命中的档案词
  reasons: string[]; // 推荐理由
};

/** 政策匹配响应（已按融合分倒序） */
export type PolicyMatchResponse = {
  items: PolicyMatchItem[];
  total: number;
};

/** 政策来源（地区/门户），含收录统计供「数据来源」页溯源 */
export type PolicySourceItem = {
  key: string;
  name: string;
  region: string;
  home_url: string; // 来源门户官网/栏目地址
  policy_count: number; // 已收录政策条数
  last_crawled_at: string | null; // 最近一次抓取时间
  item_type: string; // 机会类型（policy/competition，参赛地区选项据此过滤）
};

/** 政策来源列表响应 */
export type PolicySourceListResponse = {
  items: PolicySourceItem[];
};

// ==================== 公开政策库 API ====================

export const policyApi = {
  /** 分页浏览公开政策库（所有登录用户可访问） */
  list: (params: ListPoliciesParams = {}): Promise<PolicyListResponse> => {
    const query = new URLSearchParams();
    if (params.page) query.set("page", String(params.page));
    if (params.page_size) query.set("page_size", String(params.page_size));
    if (params.region) query.set("region", params.region);
    if (params.issuer) query.set("issuer", params.issuer);
    if (params.keyword) query.set("keyword", params.keyword);
    const qs = query.toString();
    return get<PolicyListResponse>(`/policies${qs ? `?${qs}` : ""}`);
  },

  /** 查看政策详情 */
  get: (policyId: string): Promise<PolicyDetail> => {
    return get<PolicyDetail>(`/policies/${policyId}`);
  },

  /** 列出可抓取的政策来源（地区/门户） */
  listSources: (): Promise<PolicySourceListResponse> => {
    return get<PolicySourceListResponse>("/policies/sources");
  },

  /** 后台触发指定来源抓取入库（仅 owner/admin，立即返回，约 1-2 分钟后数据可见） */
  ingest: (source = "wnd", maxPages = 3): Promise<{ source: string; max_pages: number }> => {
    return post<{ source: string; max_pages: number }>(
      `/policies/ingest?source=${encodeURIComponent(source)}&max_pages=${maxPages}`
    );
  },

  /** 按当前租户企业档案匹配公开政策候选（③匹配，即时计算） */
  match: (topK = 20): Promise<PolicyMatchResponse> => {
    return get<PolicyMatchResponse>(`/policies/match?top_k=${topK}`);
  },
};
